### PR TITLE
[DependencyInjection] Add ability to define an index for service in an injected service locator argument

### PR DESCRIPTION
**Pull request by XuruDragon**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes 
| Fixed tickets | #29203
| License       | MIT
| Doc PR        | in progress /  symfony/symfony-docs#... 

It's more or less the same thing then the PR #30257 but for a service locator argument

Add a simple way to specify an index based on a tag attribute to simplify retrieving a specific service when injecting a service locator as argument into services, but also a way to fallback to a static method on the service class.

Yaml:
```yaml
services:
  foo_service:
    class: Foo
    tags:
      - {name: foo_tag, key: foo_service}
  foo_service_tagged:
    class: Bar
    arguments:
      - !tagged_locator
          tag: 'foo_tag'
          index_by: 'key'
          default_index_method: 'static_method'
```
XML:
```xml
<?xml version="1.0" ?>

<container xmlns="http://symfony.com/schema/dic/services"
    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
    xsi:schemaLocation="http://symfony.com/schema/dic/services
    http://symfony.com/schema/dic/services/services-1.0.xsd">
  <services>
    <service id="foo" class="Foo">
        <tag name="foo_tag" key="foo_service" />
    </service>
    <service id="foo_tagged_iterator" class="Bar" public="true">
      <argument type="tagged_locator" tag="foo_tag" index-by="key" default-index-method="static_method" />
    </service>
  </services>
</container>
```
PHP:
```php
// config/services.php
use Symfony\Component\DependencyInjection\Argument\TaggedIteratorArgument;
use Symfony\Component\DependencyInjection\Argument\ServiceLocatorArgument;

$container->register(Foo::class)
        ->addTag('foo_tag', ['key' => 'foo_service']);

$container->register(App\Handler\HandlerCollection::class)
         // inject all services tagged with app.handler as first argument
         ->addArgument(new ServiceLocatorArgument(new TaggedIteratorArgument('app.handler', 'key')));
```

Usage: 
```php
// src/Handler/HandlerCollection.php
namespace App\Handler;

use Symfony\Component\DependencyInjection\ServiceLocator;

class HandlerCollection
{
     public function __construct(ServiceLocator $serviceLocator)
     {
           $foo = $serviceLocator->get('foo_service'):
     }
}
```

Tasks

* [x]  Support PHP loader/dumper
* [x]  Support YAML loader/dumper
* [x]  Support XML loader/dumper (and update XSD too)
* [x]  Add tests
* [x]  Documentation

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/PriorityTaggedServiceTraitTest.php [hunk 1]
```diff
@@ -93,6 +93,6 @@ class PriorityTaggedServiceTraitImplementation
 
     public function test($tagName, ContainerBuilder $container)
     {
-        return $this->findAndSortTaggedServices($tagName, $container);
+        return self::findAndSortTaggedServices($tagName, $container);
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/PriorityTaggedServiceTraitTest.php]: this change looks wrong to me. The method is not static

**Review by stof — CHANGES_REQUESTED** [review]

**Review by nicolas-grekas — APPROVED** [review]
(once @stof's comment is addressed)

**Comment by vudaltsov** [comment]
Is there any possibility to index by a class name?

**Comment by nicolas-grekas** [comment]
@vudaltsov sure: implement getDefautKeyName and make it return self::class.

**Comment by vudaltsov** [comment]
That will be a new static method in the interface which all tagged services implement...
What if we could add such a possibility out of the box? It's quite common to index by a class name. Because it's automatically a unique identifier and a key that is easy to use in the code.

**Comment by nicolas-grekas** [comment]
@vudaltsov the make all such classes extends some base class, and make the method of that class return static::class. That's almost what is done right now for form types.

**Comment by vudaltsov** [comment]
@nicolas-grekas , not true: https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Form/DependencyInjection/FormPass.php#L73
same for constraint validators:
https://github.com/symfony/symfony/blob/master/src/Symfony/Component/Validator/DependencyInjection/AddConstraintValidatorsPass.php#L48

**Comment by vudaltsov** [comment]
@XuruDragon  , what about `default_index_method: '::class'`?

**Comment by nicolas-grekas** [comment]
@vudaltsov thanks for facts checking :) Indexing by FQCN is now implemented as a fallback when the attribute or the method are either not defined in the tag or not found on the service.

**Comment by XuruDragon** [comment]
branch has been rebase on up-to-date master

**Comment by nicolas-grekas** [comment]
Thank you @XuruDragon.

**Pull request merged**